### PR TITLE
Fix accuracy with separate calibration layer

### PR DESCRIPTION
**Pull request by nikkolasg**

### CODE REVIEW

File: zkml/assets/scripts/CNN/cifar-cnn.py [hunk 1]
```diff
@@ -496,33 +524,482 @@ def forward(self, x):
...
+
+# num_eval_batches = 1000
+# top1, top5 = evaluate(net, criterion, testloader, neval_batches=num_eval_batches)
+# print('Evaluation accuracy on %d images, %2.2f'%(num_eval_batches, top1.avg))
+# compute_accuracy(net, testloader)
+
+# Create a subset of test data with exactly the same samples used in the JSON
+json_sample_dataset = Subset(testset, sample_indices)
+json_sample_loader = torch.utils.data.DataLoader(
+    json_sample_dataset, batch_size=batch_size, shuffle=False)
```
Comment (nicholas-mainardi) [inline zkml/assets/scripts/CNN/cifar-cnn.py:862]: Aren't these lines redundant? The same loader seems to be created below

**Review by nicholas-mainardi — COMMENTED** [review]
LGTM, moslty minor comments. Not a deep review on the padding logic for conv though, as I am not super familiar with it yet

File: zkml/src/layers/requant.rs [hunk 2]
```diff
@@ -395,3 +490,123 @@ impl RequantCtx {
...
+//            after_range: 1 << *quantization::BIT_LEN,
+//        };
+//        let res_requantft= requant.op(&res);
+//        println!("res_requant: {:?}", res_requant.get_data());
+//        println!("res_requantf: {:?}", res_requantf.get_data());
+//        println!("res_requantft: {:?}", res_requantft.get_data());
+//        //assert!(res_requant.get_data().iter().filter(|r| **r == 0 || **r == -1).collect::<Vec<_>>().len() < res_requant.get_data().len());
+//    }
+//
+//}
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/layers/requant.rs:612]: Yeah useless tests.

File: zkml/src/layers/convolution.rs [hunk 3]
```diff
@@ -85,9 +91,15 @@ impl Convolution {
...
-    pub fn op<E: ExtensionField>(&self, input: &Tensor<Element>) -> (Tensor<Element>, ConvData<E>) {
-        let (output, proving_data) = self.filter.fft_conv(input);
-        (self.add_bias(&output), proving_data)
+    /// Retrieves an element using (N, C, H, W) indexing
+    pub fn get(&self, n: usize, c: usize, h: usize, w: usize) -> T {
+        assert!(self.filter.get_shape().len() <= 4);
+
+        let (n_size, c_size, h_size, w_size) = self.filter.get4d();
+
+        assert!(n < n_size);
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/layers/convolution.rs:100]: Yeah ok i've added these assertions as you suggest. Note that in case the indices are not valid, it'll likely panic  at the last line `get_data()[index]`, so same behaviour as with the assert, but easier to catch why.

File: zkml/src/layers/convolution.rs [hunk 4]
```diff
@@ -707,65 +765,237 @@ pub fn phi_eval<E: ExtensionField>(
...
+        not_padded_shape.remove(0);
+        let mut garbage = Vec::new();
+        let mut valid = Vec::new();
+        for i in 0..fft_output.shape[0] {
+            for j in 0..fft_output.shape[1] {
+                for k in 0..fft_output.shape[2] {
+                    let index =
+                        i * fft_output.shape[1] * fft_output.shape[2] + j * fft_output.shape[2] + k;
+                    let elem = fft_output.data[index];
+                    if !(i < not_padded_shape[0]
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/layers/convolution.rs:817]: Switched to positive predicates

File: zkml/src/padding.rs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        nrows
+    );
+    assert!(si.input_shape_padded.iter().all(|d| d.is_power_of_two()));
+    if si.input_shape_padded.len() != 1 {
+        si.input_shape_padded = vec![si.input_shape_padded.iter().product()];
+        si.input_shape_og = vec![si.input_shape_og.iter().product()];
+    }
+    let mut new_cols = d.matrix.ncols_2d();
+    if d.matrix.ncols_2d() != si.input_shape_padded[0] {
+        if d.matrix.ncols_2d() < si.input_shape_padded[0] {
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/padding.rs:129]: If the number of columns in the matrix is equal to the input shape padded's first dimension (tensor  is usually flattened at this point), then there's nothing else to do. It's only if there is LESS columns than the input vector size we need to pad the matrix.

File: zkml/src/layers/hadamard.rs [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+            sumcheck_aux: VPAuxInfo::from_mle_list_dimensions(&vec![vec![
+                num_vars, num_vars, num_vars,
+            ]]),
+        }
+    }
+}
+
+pub struct HadamardProof<F: ExtensionField> {
+    sumcheck: IOPProof<F>,
+    individual_claim: Vec<F>,
```
Comment (Zyouell) [inline zkml/src/layers/hadamard.rs:36]: Since the `beta_eval` should be calculated by the verifier we shouldn't supply it.

File: zkml/src/layers/reshape.rs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    }
+}
+
+impl QuantizableOp for Reshape {
+    fn quantize(&self, _s: &ScalingFactor, _bias_s: Option<&ScalingFactor>) -> Layer<Element> {
+        Layer::Reshape(Reshape)
+    }
+}
+
+impl ProvableOp for Reshape {
```
Comment (Zyouell) [inline zkml/src/layers/reshape.rs:34]: I thought `Reshape` wasn't provable currently? If so this implementation is confusing.

File: zkml/src/onnx_parse.rs [hunk 8]
```diff
@@ -329,91 +349,60 @@ pub fn load_model<Q: Quantizer<Element>>(filepath: &str) -> Result<Model> {
         let key = item.name.to_string();
         initializers.insert(key, value);
     }
-
-    let mut layers: Vec<Layer> = Vec::new();
+    println!(
```
Comment (Zyouell) [inline zkml/src/onnx_parse.rs:352]: Can we change the `println!` to a `debug!` or make it logged?

File: zkml/src/quantization/mod.rs [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+    };
+    let max = if weight.is_negative() {
+        *weight * *min_input
+    } else {
+        *weight * *max_input
+    };
+    (min, max)
+}
+
+pub trait MinMax {
```
Comment (Zyouell) [inline zkml/src/quantization/mod.rs:212]: I think this traits functionality got absorbed into `Number` if so can we get rid of it?

File: zkml/src/tensor.rs [hunk 10]
```diff
@@ -23,6 +29,101 @@ use crate::{
...
+        Element::random(rng).to_field()
+    }
+    fn absolute_value(&self) -> Self {
+        *self
+    }
+    fn compare(&self, other: &Self) -> Ordering {
+        self.cmp(other)
+    }
+
+    fn is_negative(&self) -> bool {
```
Comment (Zyouell) [inline zkml/src/tensor.rs:122]: Technically this could be useful when converting between element and field as we always do a "negative" check by seeing if we are greater than half the field modulus. It may make the rest of the code cleaner to implement that here and then use it elsewhere.

File: zkml/src/tensor.rs [hunk 11]
```diff
@@ -510,11 +649,23 @@ where
...
-    T: std::default::Default,
+    T: Number,
 {
+    pub fn min(&self) -> T {
+        self.data.iter().fold(T::MAX, |min, x| min.cmp_min(x))
+    }
+    pub fn max(&self) -> T {
+        self.data.iter().fold(T::MIN, |max, x| max.cmp_max(x))
+    }
+    pub fn reshape(mut self, new_shape: Vec<usize>) -> Tensor<T> {
```
Comment (Zyouell) [inline zkml/src/tensor.rs:660]: This function should really check that the new shape is valid, in that the total number of elements in the tensors don't change

File: zkml/src/layers/convolution.rs [hunk 12]
```diff
@@ -69,12 +75,12 @@ pub struct ConvProof<E: ExtensionField> {
     bias_claim: E,
 }
 
-impl Convolution {
-    pub fn new(filter: Tensor<Element>, bias: Tensor<Element>) -> Self {
+impl<T: Number> Convolution<T> {
+    pub fn new(filter: Tensor<T>, bias: Tensor<T>) -> Self {
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/layers/convolution.rs:79]: I didn't change this but ok - wrote the issue here https://github.com/Lagrange-Labs/deep-prove/issues/106 - i would like to tacke this in a separate PR since there are many places that still rely on assert, it's a full blown PR to do that.

[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/layers/convolution.rs:79]: Added the check on length 4 tho

File: zkml/src/layers/convolution.rs [hunk 13]
```diff
@@ -116,59 +128,105 @@ impl Convolution {
...
-            }
+    /// Quantizes the filter and the bias. Note the weights are not yet FFT'd, that happens with new_conv at the padding step
+    /// since the FFT is also making the filter shape == input shape.
+    /// TODO: refactor new_conv to be in convolution.rs and more efficient than cloning
+    /// It uses a custom scaling factor `bias_s` for the bias, if provided,
+    /// otherwise the same scaling factor of the weights (i.e., `s`) is used
+    pub fn quantize(
+        self,
+        s: &ScalingFactor,
+        bias_s: Option<&ScalingFactor>,
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/layers/convolution.rs:150]: That's coming from how PyTorch is doing things where they create a manual scaling factor using the scaling factor of the filter - @nicholas-mainardi can explain more - it was a test he added to reflect exactly that same accuracy of PyTorch. But it's a good question to whether we should drop it or not.

File: zkml/src/layers/convolution.rs [hunk 14]
```diff
@@ -116,59 +128,105 @@ impl Convolution {
...
+    pub fn float_op(&self, input: &Tensor<f32>) -> Tensor<f32> {
+        input.conv2d(&self.filter, &self.bias, 1)
+    }
+}
+
+impl Convolution<Element> {
+    pub fn op<E: ExtensionField>(&self, input: &Tensor<Element>) -> (Tensor<Element>, ConvData<E>) {
+        let (output, proving_data) = self.filter.fft_conv(input);
+        (self.add_bias(&output), proving_data)
+        // create a new tensor that clears out the garbage values in the output, e.g. set all to 0
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/layers/convolution.rs:183]: I mean they apply but too soon - it's in next PR ;) Removed.

File: zkml/src/layers/convolution.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/layers/convolution.rs:150]: Removing the option.

File: zkml/src/layers/hadamard.rs [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+
+#[allow(unused)]
+pub fn verify<F: ExtensionField, T: Transcript<F>>(
+    ctx: &HadamardCtx<F>,
+    transcript: &mut T,
+    proof: &HadamardProof<F>,
+    output_claim: Claim<F>,
+    expected_v2_eval: F,
+) -> Result<Claim<F>> {
+    let _subclaim = IOPVerifierState::<F>::verify(
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/layers/hadamard.rs:112]: Good catch !

File: zkml/src/layers/mod.rs [hunk 16]
```diff
@@ -182,14 +163,160 @@ impl Layer {
...
+                    d.mean().unwrap(),
+                    d.std_dev().unwrap()
+                );
+            }
+        }
+        Ok(output)
+    }
+}
+
+use hex;
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/layers/mod.rs:309]: Testing but it turned out to be useless. Removing it.

File: zkml/src/layers/requant.rs [hunk 17]
```diff
@@ -395,3 +490,123 @@ impl RequantCtx {
         Ok(Claim { point, eval })
     }
 }
+
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/layers/requant.rs]: That doesn't really test anything, it's just for printing. It's a test that no one uses I think - unless you do ?
What we should do is implement [the testing strategy laid out in this blog post](https://leimao.github.io/article/Neural-Networks-Quantization/) tho. This would be a true test.

File: zkml/src/onnx_parse.rs [hunk 18]
```diff
@@ -541,91 +539,81 @@ fn extract_tensor_f32_data(
     let tensor_vec = node
         .input
         .iter()
-        .filter(|x| x.contains(weight_or_bias))
+        .filter(|x| is_good_selector(x))
         .filter_map(|key| initializers.get(key).cloned())
         .collect_vec();
 
+    println!(
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/onnx_parse.rs:546]: Put to trace

File: zkml/src/quantization/metadata.rs [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+        self.input_scaling = Some(input_scaling);
+    }
+
+    pub fn set_layers_scaling(&mut self, layer_id: usize, output_scaling: ScalingFactor) {
+        self.layers_scaling.insert(layer_id, output_scaling);
+    }
+
+    pub fn build(self) -> ModelMetadata {
+        ModelMetadata {
+            input: self.input_scaling.unwrap(),
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/quantization/metadata.rs:52]: We _have_ to have something : reason is simple, this metadata only gets returned by the strategy that takes in the float model. But i'm changing it such that we dont need the unwrap anymore, by requiring the input scaling factor in the constructor.

File: zkml/src/quantization/mod.rs [hunk 20]
```diff
@@ -0,0 +1,319 @@
...
+            min: min,
+            max: max,
+            quantized_domain: quantized_domain.unwrap_or((*MIN, *MAX)),
+        }
+    }
+    // Initialize a scaling factor in such a way that `self.scale()` is equal to the `scale` value
+    // provided as input.
+    pub(crate) fn from_scale(scale: f32, quantized_domain: Option<(Element, Element)>) -> Self {
+        let (min_quantized, max_quantized) = quantized_domain.clone().unwrap_or((*MIN, *MAX));
+        let max = scale / 2.0 * (max_quantized - min_quantized) as f32;
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/quantization/mod.rs:72]: `S = (r_max - r_min) / (q_max - q_min)` since we do symmetric scaling, it turns out `S = 2 * max_abs_r / (q_max - q_min)` so `max_abs_r = S * (q_max - q_min) / 2`

File: zkml/src/quantization/mod.rs [hunk 21]
```diff
@@ -0,0 +1,319 @@
...
+    }
+
+    /// Take a floating point number and quantize it to an BIT_LEN-bit integer
+    /// S = (a - (-a)) / (2^{BIT_LEN-1}- (-2^{BIT_LEN-1})) = 2a / 2^BIT_LEN
+    pub fn quantize(&self, value: &f32) -> Element {
+        // assert!(
+        //    *value >= -1.0 && *value <= 1.0,
+        //    "Input value must be between -1.0 and 1.0"
+        //);
+        let zero_point = 0;
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/quantization/mod.rs:106]: idk - @sshravan put it there first and i kinda liked it that it's explicit that we are using zero point. The cost of allocating a 0-valued variable should be null so it doesn't make any difference.

[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/quantization/mod.rs:106]: But making it a constant now so the compilator will know  for sure it can be optimized away.

File: zkml/src/quantization/mod.rs [hunk 22]
```diff
@@ -0,0 +1,319 @@
...
+            e
+        } else {
+            // That means he's a negative number - so take the diff with the modulus and recenter around 0
+            let diff = <F::BaseField as SmallField>::MODULUS_U64 - e as u64;
+            -(diff as Element)
+        }
+    }
+}
+
+impl<F: ExtensionField> Fieldizer<F> for u8 {
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/quantization/mod.rs:173]: Nope - removed

File: zkml/src/quantization/strategy.rs [hunk 23]
```diff
@@ -0,0 +1,380 @@
...
+                .get(&layer_index)
+                .expect(&format!("No data for layer {:?}", layer_index))
+                .clone(),
+        );
+        let min = d.percentile(5) as f32;
+        let max = d.percentile(95) as f32;
+        assert!(min <= max);
+        //(min, max)
+        (d.min() as f32, d.max() as f32)
+        // let mean = d.mean().unwrap();
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/quantization/strategy.rs:264]: Actually unclear - let's see when we start dealing with inputs that are OUT OF RANGE and see how it behaves - maybe it's best to take min and max from distribution rather than actual min and max..

File: zkml/src/tensor.rs [hunk 24]
```diff
@@ -451,6 +576,20 @@ impl<T> Tensor<T> {
     pub fn get_data(&self) -> &[T] {
         &self.data
     }
+
+    pub fn kx(&self) -> usize {
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/tensor.rs:580]: https://github.com/Lagrange-Labs/deep-prove/issues/107 - let's fix that properly, there's a lot of work to get rid fully of conv related notion in tensors.rs - not gonna do that here.

File: zkml/src/quantization/mod.rs (hunk 9 above)
Comment (nikkolasg) [reply zkml/src/quantization/mod.rs:212]: Not quite the same tho. I dont want to overload the Number trait but there is no `to _f32` for example.

File: zkml/src/tensor.rs (hunk 10 above)
Comment (nikkolasg) [reply zkml/src/tensor.rs:122]: Right now since we _dont_ use this functionality i didn't care implementing it. If we do want to use, then that's something else. Not sure why it's relevant now to implement it if unused.

**Review by Zyouell — APPROVED** [review]
